### ISSUE TITLE
printing Hamiltonians with general operations

### ISSUE THREAD

**Issue by albi3ro** [PennyLaneAI/pennylane#982]
[`vqe.Hamiltonian.__str__`](https://github.com/PennyLaneAI/pennylane/blob/1e15072e71ce83e455d454123cc5492781cc1e2a/pennylane/vqe/vqe.py#L160) cannot handle operations of type `pennylane.ops.qubit.Hermitian`. Currently, the function only works with operations of type `Tensor` and `Observable`.

We can see such a case in this example from the testing suite:

```python
H_TWO_QUBITS = np.array(
    [[0.5, 1.0j, 0.0, -3j], [-1.0j, -1.1, 0.0, -0.1],
     [0.0, 0.0, -0.9, 12.0], [3j, -0.1, 12.0, 0.0]]
)

H = qml.vqe.Hamiltonian((1.0, ), (qml.Hermitian(H_TWO_QUBITS, [0,1]), ) )

print(H)
```
 
We should also add tests of the `vqe.Hamiltonian.__str__` function so issues like this are caught.

**Comment by albi3ro** [comment]
So the problem wasn't it not being a `Tensor` or `Observable`, but not something specified in [`OBS_MAP`](https://github.com/PennyLaneAI/pennylane/blob/1e15072e71ce83e455d454123cc5492781cc1e2a/pennylane/vqe/vqe.py#L27)

```python
OBS_MAP = {"PauliX": "X", "PauliY": "Y", "PauliZ": "Z", "Hadamard": "H", "Identity": "I"}
```

The question now is, how do we want to print an arbitrary matrix?

I propose

```
(1.0) [a01]
---
a = [[0.5, 1.0j, 0.0, -3j], [-1.0j, -1.1, 0.0, -0.1],
     [0.0, 0.0, -0.9, 12.0], [3j, -0.1, 12.0, 0.0]]
```

I am unsure whether or not we need to print the matrices at the bottom

Each time we get another non-enumerated operation, it gets the next letter.

**Comment by antalszava** [comment]
Oh, that's a good catch! :slightly_smiling_face: Think that having `(1.0) [Herm0]` or `(1.0) [Herm1]` could work (though this is very much subjective).

If the matrix is not printed, the user could inspect it through the `Hamiltonian.terms` attribute.

```python
In [13]: a = qml.vqe.Hamiltonian((1, ), (qml.Hermitian(np.array([1,2]), wires=1), ))                                                              

In [14]: a                                                                                                                                        
Out[14]: <pennylane.vqe.vqe.Hamiltonian at 0x7f348953de10>

In [15]: a.terms                                                                                                                                  
Out[15]: ((1,), (Hermitian(array([1, 2]), wires=[1]),))
```

**Comment by albi3ro** [comment]
If the Hamiltonian has more than one non-specified operation, we need a way to differentiate between them.  Sticking a number after the operator is how we designated what wires it acts on, ex `a01` would act on wires `0` and `1`.  Therefore we can't necessarily use numbers to differentiate between different operators.  This is why I thought different letters.

We might also be able to denote it by `(1.0) [Herm0-0]` or `(1.0) [op0-0]`. 

If we go this way, we could have the number denote its index in the operator array.

Bonus idea:
If the same operator (same matrix) gets applied again to different wires, can we keep the same label?

**Comment by antalszava** [comment]
Got it! Think that works well, perhaps the identifiers (e.g., `a`) could be more descriptive (one of the alternatives `Herm`, `op` or perhaps `mx` might be slightly better).

----

*tl;dr:* Overall, I think it's okay to keep printing out information as simple and intuitive as possible, so I think it could be okay to encode less information into the string representation and have users depend on inspecting objects.

----

> We might also be able to denote it by (1.0) [Herm0:0] or (1.0) [op0:0].

Would the number left to `:` refer to the index in the operator array and the characters to the right encode the wires?

One thing to note here is that there could be custom wire labels too. E.g., `(1.0) [Herm0:wire1wire2]`.

This could make the string representation less intuitive with more exquisite custom wire labels. For example having `wires=['wire1', 1, 'c']` would produce `(1.0) [Herm0:wire11c]`, where encoding the wires leads to ambiguity. For cases like these, I'd be tempted to consider if it's worth encoding information about the wires.

----

> If the same operator (same matrix) gets applied again to different wires, can we keep the same label?

That's a fair question! Two separate cases come to mind here:

a) the same operator object is used multiple times to construct the hamiltonian

```python
In [5]: herm = qml.Hermitian(np.array([0,1]), 0)                                                                                                  

In [6]: qml.vqe.Hamiltonian((0.5,0.5, ), (herm, herm, ))                                                                                          
Out[6]: <pennylane.vqe.vqe.Hamiltonian at 0x7f34daf48750>

```
b) the same matrix is used to create the same object multiple times when constructing the hamiltonian
```python
In [7]: mx = np.array([0,1])                                                                                                                      

In [8]: qml.vqe.Hamiltonian((0.5,0.5, ), (qml.Hermitian(mx,0), qml.Hermitian(mx,0), ))                                                            
Out[8]: <pennylane.vqe.vqe.Hamiltonian at 0x7f34daf46550>
```
Think it might be worth checking for b), although this might become complex as for each hermitian term, we'd need to keep track of matrices and labels. Then each term needs to be checked and labelled correctly. Not completely sure if it's worth the complexity that is introduced for labelling (might be wrong though).
